plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/p5edges/orthogonal/HyperEdgeCycleDetector.java

```diff
@@ -78,25 +86,56 @@ private HyperEdgeCycleDetector() {
      * complete, all segments are marked from {@code -1} to {@code -segments.size()}.
      */
     private static void initialize(final List<HyperEdgeSegment> segments, final List<HyperEdgeSegment> sources,
-            final List<HyperEdgeSegment> sinks) {
+            final List<HyperEdgeSegment> sinks, final boolean criticalOnly) {
         
         int nextMark = -1;
         for (HyperEdgeSegment segment : segments) {
             segment.mark = nextMark--;
+            
             int inWeight = 0;
             int outWeight = 0;
+            
+            int criticalInWeight = 0;
```

Comment:
Maybe `int criticalInWeight = segment.getOutgoingSegmentDependencies().stream().filter(d -> d.getType() == CRITICAL).sumInt()`?

Same without the filter for `inWeight`?